dart/gui/osg/Viewer.cpp

```diff
@@ -831,6 +852,64 @@ const ::osg::ref_ptr<::osg::Group>& Viewer::getRootGroup() const
   return mRootGroup;
 }
 
+//==============================================================================
+const ::osg::ref_ptr<::osg::Group>& Viewer::getPhysicsGroup() const
+{
+  return mPhysicsGroup;
+}
+
+//==============================================================================
+bool Viewer::isShadowed() const
+{
+  return mShadowed;
+}
+
+//==============================================================================
+void Viewer::enableShadows(bool _enable, ::osg::ref_ptr<osgShadow::ShadowTechnique> shadowTechnique)
+{
+  if(!mPhysicsGroup) {
+    // Flags for shadowing; maybe this needs to be global?
+    constexpr int ReceivesShadowTraversalMask = 0x2;
+    constexpr int CastsShadowTraversalMask = 0x1;
+
+    // Setup shadows
+    // Create one ShadowedScene for each light
+    ::osg::ref_ptr<osgShadow::ShadowedScene> shadowedScene = new osgShadow::ShadowedScene;
+    shadowedScene->setReceivesShadowTraversalMask(ReceivesShadowTraversalMask);
+    shadowedScene->setCastsShadowTraversalMask(CastsShadowTraversalMask);
+
+    // set the physics group
+    mPhysicsGroup = shadowedScene.get();
+    mPhysicsGroup->getOrCreateStateSet();
+
+    setupDefaultLights();
+  }
+
+  if(_enable)
+    setShadowTechnique(shadowTechnique);
+  else
+    static_cast<osgShadow::ShadowedScene*>(mPhysicsGroup.get())->setShadowTechnique(0);
+
+  mShadowed = _enable;
+}
+
+void Viewer::setShadowTechnique(::osg::ref_ptr<osgShadow::ShadowTechnique> shadowTechnique) {
```

Comment:
I see now that if the user passes a `nullptr`, we replace it with a default `osgShadow::ShadowMap`.

I think that's a bit opaque, so what if we instead make an `osg::ref_ptr<osgShadow::ShadowMap>` instance the default value for these functions?

Reply:
> I think that's a bit opaque, so what if we instead make an osg::ref_ptr<osgShadow::ShadowMap> instance the default value for these function arguments?

Yes we can do this..

> I'm further realizing that the initialization of the ShadowMap is using one of the member variables of the Viewer class, mLight1.

That's why I exposed the `LightSources` so that users can define their own `ShadowTechnique`s and be able to choose the light source that will cast the shadows. We can't do it generically because not all of the techniques have the `setLight()` functionality..